File: Dynamic programming/55_JumpGame.cs
// 55. Jump Game
// Tags: dynamic programming, greedy
//

public class Solution 
{
    public bool CanJump(int[] nums) 
    {
        // Edge case.
        if(nums == null || nums.Length == 0)
            return false;

        // Optimizations.
        if(nums.Length == 1)
            return true;
        else if(nums[0] == 0)
            return false;

        int dist = 1;
        // Greedy. Count backwards to see if it is possible to reach the end
        // of the array.
        // Note we are not interested in the last index since it's value is
        // allowed to be anything.
        for(int i = nums.Length - 2; i >= 0; i--)
        {
            if(nums[i] >= dist)
                dist = 0;

            dist++;
        }

        return (dist <= nums[0]) ? true : false;
    }
}

File: Dynamic programming/55_JumpGame.rs
// 55. Jump Game
// Tags: dynamic programming, greedy
//

impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {

        // Edge case.
        if nums.is_empty() || nums.len() == 0 {
            return false;
        }

        // Optimizations.
        if nums.len() == 1 {
            return true;
        }
        else if nums[0] == 0 {
            return false;
        }

        let mut dist: i32 = 1;
        let mut i: usize = nums.len() - 1;

        while i > 0 {
            i -= 1;

            if nums[i] >= dist {
                dist = 0;
            }
            dist += 1;
        }

        if dist <= nums[0] { 
            return true;
        }
        false
    }
}